### PR TITLE
feat(ilp): ilp time based commit

### PR DESCRIPTION
**Pull request by glasstiger**
New commit strategy for ILP.
- Num of rows based: After a configured number of uncommitted rows a commit is issued, this parameter is set per table.
- Time interval based: A commit is issued periodically, the time interval is the fraction of the commit lag configured for the table. The fraction is set in the config, if not set it is defaulted to 0.5. If there is no commit lag set for the table or the fraction is set to 0.0 in the config a default commit interval will be used instead. This default interval also configurable.
- Idle time based: When the table goes idle and the writer is released, a commit is issued.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 56 / 61 (91.80%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|20|24|83.33%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpWriterJob.java|16|17|94.12%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DefaultLineTcpReceiverConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 57 / 61 (93.44%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|21|24|87.50%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpWriterJob.java|16|17|94.12%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DefaultLineTcpReceiverConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 56 / 59 (94.92%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|20|23|86.96%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpWriterJob.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DefaultLineTcpReceiverConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 55 / 59 (93.22%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|19|23|82.61%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpWriterJob.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DefaultLineTcpReceiverConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 55 / 59 (93.22%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|19|23|82.61%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpWriterJob.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DefaultLineTcpReceiverConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 55 / 59 (93.22%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|19|23|82.61%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpWriterJob.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DefaultLineTcpReceiverConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 55 / 59 (93.22%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|19|23|82.61%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpWriterJob.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DefaultLineTcpReceiverConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|1|1|100.00%|

File: core/src/main/java/io/questdb/cairo/TableWriter.java [hunk 1]
```diff
@@ -52,6 +52,7 @@
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.LongConsumer;
 
+import static io.questdb.PropServerConfiguration.*;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/TableWriter.java:55]: can we format this file please, method order is custom, unused imports

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpWriterJob.java [hunk 2]
```diff
@@ -87,34 +87,34 @@ public boolean run(int workerId) {
...
-                    // if the queue is still empty we will come back and other tables will have their chance for a commit too
-                    // obviously tables at the end of the list (assignedTables) are disadvantaged but commits are
-                    // still issued automatically after we reached the max number of uncommitted rows and eventually
-                    // all tables in the list will get their commit after ingestion stops or slows down
-                    // the heap based solution mentioned above will eliminate this problem completely
-                    return true;
+                // the heap based solution mentioned above will eliminate the minimum search
+                // we could just process the min element of the heap until we hit the first commit
+                // time greater than millis and that will be our nextCommitTime
+                long tableNextCommitTime = assignedTables.getQuick(n).checkIfTableNeedsCommit(millis);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpWriterJob.java:109]: checkIfTableNeedsCommit() is a confusing name, it does do commit whereas name doesn't lead you to believe it does. Can we rename ?

File: core/src/main/java/io/questdb/cutlass/line/tcp/TableUpdateDetails.java [hunk 3]
```diff
@@ -196,63 +194,61 @@ TableWriter getWriter() {
...
+        writer.commitWithLag();
         // Tick after commit.
         writer.tick(false);
-        return true;
     }
 
-    boolean commitIfTimeoutElapsed(long millis) {
-        if (millis - lastMeasurementMillis < noAppendInterval || millis - lastCommitMillis < commitTimeout) {
-            return false;
+    long checkIfTableNeedsCommit(long millis) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cutlass/line/tcp/TableUpdateDetails.java:213]: nit: could we rename `millis` to `wallClockMillis` in this method and everywhere its used?

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpWriterJob.java (hunk 2 above)
Comment (glasstiger) [reply core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpWriterJob.java]: done

File: core/src/main/java/io/questdb/cutlass/line/tcp/TableUpdateDetails.java (hunk 3 above)
Comment (glasstiger) [reply core/src/main/java/io/questdb/cutlass/line/tcp/TableUpdateDetails.java]: done

**Review by bluestreak01 — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 56 / 60 (93.33%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|19|23|82.61%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpWriterJob.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DefaultLineTcpReceiverConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|1|1|100.00%|

**Pull request merged**